be/src/io/fs/hdfs_file_system.h

```diff
@@ -0,0 +1,159 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <gen_cpp/PlanNodes_types.h>
+#include <hdfs/hdfs.h>
+
+#include "common/status.h"
+#include "io/fs/remote_file_system.h"
+#include "io/hdfs_file_reader.h"
+namespace doris {
+
+namespace io {
+
+class HdfsFileSystemHandle {
+public:
+    HdfsFileSystemHandle(hdfsFS fs, bool cached)
+            : hdfs_fs(fs), from_cache(cached), _ref_cnt(0), _last_access_time(0), _invalid(false) {}
+
+    ~HdfsFileSystemHandle() {
+        DCHECK(_ref_cnt == 0);
+        if (hdfs_fs != nullptr) {
+            // Even if there is an error, the resources associated with the hdfsFS will be freed.
+            hdfsDisconnect(hdfs_fs);
+        }
+        hdfs_fs = nullptr;
+    }
+
+    int64_t last_access_time() { return _last_access_time; }
+
+    void inc_ref() {
+        _ref_cnt++;
+        _last_access_time = _now();
+    }
+
+    void dec_ref() {
+        _ref_cnt--;
+        _last_access_time = _now();
+    }
+
+    int ref_cnt() { return _ref_cnt; }
+
+    bool invalid() { return _invalid; }
+
+    void set_invalid() { _invalid = true; }
+
+    hdfsFS hdfs_fs;
+    // When cache is full, and all handlers are in use, HdfsFileSystemCache will return an uncached handler.
+    // Client should delete the handler in such case.
+    const bool from_cache;
+
+private:
+    // the number of referenced client
+    std::atomic<int> _ref_cnt;
+    // HdfsFileSystemCache try to remove the oldest handler when the cache is full
+    std::atomic<uint64_t> _last_access_time;
+    // Client will set invalid if error thrown, and HdfsFileSystemCache will not reuse this handler
+    std::atomic<bool> _invalid;
+
+    uint64_t _now() {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(
+                       std::chrono::system_clock::now().time_since_epoch())
+                .count();
+    }
+};
+
+// Cache for HdfsFileSystemHandle
+class HdfsFileSystemCache {
+public:
```

Comment:
If `HdfsFileSystemCache` won't be used outside hdfs file system, had better define this class in .cpp